src/reformed.js

```diff
@@ -27,39 +35,98 @@ const makeWrapper = (middleware) => (WrappedComponent) => {
       }))
     }
 
+    initInputFlags = (name) => {
+      if (!this.inputFlags[name]) {
+        this.setInputFlags(name, {
+          dirty: false,
+          pristine: true,
+          touched: false,
+          untouched: true
+        })
+      }
+    }
+
+    setInputFlags = (name, flags) => {
+      const merged = assign({}, this.inputFlags[name], flags)
+      const inverse = {
+        pristine: !merged.dirty,
+        untouched: !merged.touched
+      }
+
+      this.inputFlags[name] = assign({}, merged, inverse)
+      return this.inputFlags
+    }
+
     // This, of course, does not handle all possible inputs. In such cases,
     // you should just use `setProperty` or `setModel`. Or, better yet,
     // extend `reformed` to supply the bindings that match your needs.
     bindToChangeEvent = (e) => {
       const { name, type, value } = e.target
+      const newFlags = { dirty: true }
 
       if (type === 'checkbox') {
         const oldCheckboxValue = this.state.model[name] || []
         const newCheckboxValue = e.target.checked
           ? oldCheckboxValue.concat(value)
           : oldCheckboxValue.filter(v => v !== value)
 
+        newFlags.touched = true
         this.setProperty(name, newCheckboxValue)
       } else {
         this.setProperty(name, value)
       }
+
+      this.setInputFlags(name, newFlags)
     }
 
     bindInput = (name) => {
+      this.initInputFlags(name)
       return {
         name,
         value: this.state.model[name] || '',
-        onChange: this.bindToChangeEvent,
+        onChange: this.onInputEvent,
+        onBlur: this.onInputEvent,
+        onFocus: this.onInputEvent
       }
     }
 
+    bindCheckbox = (name, value) => {
+      this.initInputFlags(name)
+      return {
+        name,
+        checked: contains(value, this.state.model[name]),
+        onChange: this.onInputEvent,
+        onBlur: this.onInputEvent,
+        onFocus: this.onInputEvent
+      }
+    }
+
+    onInputEvent = (e) => {
+      const { target, type } = e
+      const { name } = target
+      const lastInputEvent = { name, target, type }
+
+      this.setState({ lastInputEvent })
+      if (type === 'change') this.bindToChangeEvent(e)
+      if (type === 'blur') this.setInputFlags(name, { touched: true })
+    }
+
+    componentDidMount = () => {
+      // Once downstream inputs are bound and `inputFlags`
+      // has been set, force an update
+      this.forceUpdate()
```

Comment:
Yeah, it's not necessary if `reformed` is used in conjunction with `syncWith`, because populating the state essentially triggers a re-render which picks up the inputFlags. Agree it would be better to not have to do that, but I haven't thought of a better way.